### PR TITLE
ASoC: SOF: Intel: better split between codec and controller setup

### PR DESCRIPTION
**Pull request by plbossart**
First draft to cluster all codec related stuff in hda-codec, so that we can deal with nocodec setups without having to recompile.

### CODE REVIEW

File: sound/soc/sof/intel/hda-stream.c [hunk 1]
```diff
@@ -785,23 +782,9 @@ irqreturn_t hda_dsp_stream_threaded_handler(int irq, void *context)
 		active = hda_dsp_stream_check(bus, status);
 
 		/* check and clear RIRB interrupt */
-#if IS_ENABLED(CONFIG_SND_SOC_SOF_HDA)
 		if (status & AZX_INT_CTRL_EN) {
```
Comment (RanderWang) [inline sound/soc/sof/intel/hda-stream.c:785]: check-patch always tell me to remove { and } for one line if() xxxx.

File: sound/soc/sof/intel/hda-bus.c [hunk 2]
```diff
@@ -70,11 +66,13 @@ static const struct hdac_bus_ops bus_core_ops = {
...
  * This can be used for both with/without hda link support.
  */
-void sof_hda_bus_init(struct hdac_bus *bus, struct device *dev)
+void sof_hda_bus_init(struct snd_sof_dev *sdev, struct device *dev)
 {
-#if IS_ENABLED(CONFIG_SND_SOC_SOF_HDA)
+	struct hdac_bus *bus = sof_to_bus(sdev);
+
+#if IS_ENABLED(CONFIG_SND_SOC_SOF_HDA_AUDIO_CODEC)
 	snd_hdac_ext_bus_init(bus, dev, &bus_core_ops, sof_hda_ext_ops);
```
Comment (RanderWang) [inline sound/soc/sof/intel/hda-bus.c:74]: do we need it for CONFIG_SND_SOC_SOF_HDA_AUDIO_HDMI ?

File: sound/soc/sof/intel/hda-mlink.c [hunk 3]
```diff
@@ -34,6 +34,20 @@ void hda_bus_ml_get_capabilities(struct hdac_bus *bus)
...
+	struct hdac_ext_link *hlink;
+
+	if (!bus->mlcap)
+		return;
+
+	while (!list_empty(&bus->hlink_list)) {
+		hlink = list_first_entry(&bus->hlink_list, struct hdac_ext_link, list);
+		list_del(&hlink->list);
+		kfree(hlink);
+	}
```
Comment (RanderWang) [inline sound/soc/sof/intel/hda-mlink.c:48]: another way in Linux Kernel Example Programming

````c     
   /* Go through the list and free the memory. */
        struct my_list *cursor, *temp;

        list_for_each_entry_safe(cursor, temp, &Head_Node, list) {
            list_del(&cursor->list);
            kfree(cursor);
        }
````

File: sound/soc/sof/intel/hda-bus.c (hunk 2 above)
Comment (plbossart) [reply sound/soc/sof/intel/hda-bus.c:74]: This config does not exist. We use the same support for external and display audio codecs now.

File: sound/soc/sof/intel/hda-stream.c (hunk 1 above)
Comment (plbossart) [reply sound/soc/sof/intel/hda-stream.c:785]: yes indeed.

**Review by RanderWang — APPROVED** [review]

**Review by bardliao — APPROVED** [review]

**Review by ujfalusi — APPROVED** [review]
Long series with lots of moving... I'm not sure if I understand all of it, but the end result works well and the code is much better organized.

File: sound/soc/sof/intel/hda-mlink.c (hunk 3 above)
Comment (plbossart) [reply sound/soc/sof/intel/hda-mlink.c:48]: I'll add this as a fixup, thanks @RanderWang

File: sound/soc/sof/intel/hda.h [hunk 4]
```diff
@@ -742,6 +742,18 @@ static inline int hda_codec_i915_exit(struct snd_sof_dev *sdev) { return 0; }
 
 #endif
 
+#if IS_ENABLED(CONFIG_SND_SOC_SOF_HDA)
```
Comment (ranj063) [inline sound/soc/sof/intel/hda.h:764]: perhaps we can introduce a new header file as well @plbossart?

File: sound/soc/sof/intel/hda-bus.c [hunk 5]
```diff
@@ -89,5 +87,14 @@ void sof_hda_bus_init(struct hdac_bus *bus, struct device *dev)
...
+#endif /* CONFIG_SND_SOC_SOF_HDA_AUDIO_CODEC */
+}
+
+void sof_hda_bus_exit(struct snd_sof_dev *sdev)
+{
+#if IS_ENABLED(CONFIG_SND_SOC_SOF_HDA_AUDIO_CODEC)
+	struct hdac_bus *bus = sof_to_bus(sdev);
+
+	snd_hdac_ext_bus_exit(bus);
+#endif
```
Comment (ranj063) [inline sound/soc/sof/intel/hda-bus.c:99]: for the sake of consistency dont we need the #else portion for bus exit too?

File: sound/soc/sof/intel/hda-dsp.c [hunk 6]
```diff
@@ -726,33 +722,23 @@ static int hda_resume(struct snd_sof_dev *sdev, bool runtime_resume)
...
-					dev_err(sdev->dev,
-						"error %d in %s: failed to power up links",
-						ret, __func__);
-					return ret;
-				}
-			}
+		ret = hda_bus_ml_resume(bus);
+		if (ret < 0) {
+			dev_err(sdev->dev,
+				"error %d in %s: failed to power up links",
```
Comment (ranj063) [inline sound/soc/sof/intel/hda-dsp.c:738]: we dont need the __func__ anymore right?

**Review by ranj063 — APPROVED** [review]
Looks good @plbossart. I have a few minor comments but maybe we can address those as fixups or follow ups PRs

File: sound/soc/sof/intel/hda-dsp.c (hunk 6 above)
Comment (plbossart) [reply sound/soc/sof/intel/hda-dsp.c:738]: yes, will add fixups for the ml helpers and callers.

File: sound/soc/sof/intel/hda-bus.c (hunk 5 above)
Comment (plbossart) [reply sound/soc/sof/intel/hda-bus.c:99]: yes that part is odd indeed.

````
void snd_hdac_ext_bus_exit(struct hdac_bus *bus)
{
	snd_hdac_bus_exit(bus);
	WARN_ON(!list_empty(&bus->hlink_list));
}
EXPORT_SYMBOL_GPL(snd_hdac_ext_bus_exit);
````

but if SOF_HDA_AUDIO_CODEC is not defined we don't use this at all.

We'll have to revisit this later I guess.

File: sound/soc/sof/intel/hda.h (hunk 4 above)
Comment (plbossart) [reply sound/soc/sof/intel/hda.h:764]: not following the suggestion @ranj063 are you saying hda.h is becoming the dumping ground for everything and we might want to split in smaller headers?

**Comment by plbossart** [comment]
Since we have IPC4 tests let's redo the PR tests.

**Comment by plbossart** [comment]
SOFCI TEST

File: sound/soc/sof/intel/hda.h (hunk 4 above)
Comment (ranj063) [reply sound/soc/sof/intel/hda.h:764]: no no. I'm only suggesting a new header for hda-ml

Comment (plbossart) [reply sound/soc/sof/intel/hda.h:764]: ok, let's see what we can do.

**Comment by plbossart** [comment]
https://sof-ci.01.org/linuxpr/PR3884/build1149/devicetest/ show issues only in IPC4 mode, which is completely orthogonal to this PR.

Merging.

**Pull request merged**